client/app/components/EditParameterSettingsDialog.jsx

```diff
@@ -0,0 +1,204 @@
+import { includes, words, capitalize, clone, isNull } from 'lodash';
+import React, { useState, useEffect } from 'react';
+import PropTypes from 'prop-types';
+import Modal from 'antd/lib/modal';
+import Form from 'antd/lib/form';
+import Checkbox from 'antd/lib/checkbox';
+import Select from 'antd/lib/select';
+import Input from 'antd/lib/input';
+import Divider from 'antd/lib/divider';
+import { wrap as wrapDialog, DialogPropType } from '@/components/DialogWrapper';
+import { Query } from '@/services/query';
+
+const { Option } = Select;
+const formItemProps = { labelCol: { span: 6 }, wrapperCol: { span: 16 } };
+
+function useQuerySearch() {
+  const [searchTerm, setSearchTerm] = useState(); // search term
+  const [searchResults, setSearchResults] = useState([]); // search results
+
+  useEffect(() => {
+    if (!searchTerm || searchTerm.length < 3) {
+      return;
+    }
+
+    Query.query({ q: searchTerm }, ({ results }) => {
+      setSearchResults(results);
+    });
+  }, [searchTerm]); // search when term changes
+
+  return [searchResults, setSearchResults, setSearchTerm];
+}
+
+function useParamSave(onConfirm, onClose) {
+  const [saving, setSaving] = useState(false);
+
+  const save = (param) => {
+    setSaving(true);
+    onConfirm(param)
+      .then(() => {
+        onClose(param);
+      }).finally(() => {
+        setSaving(false);
+      });
+  };
+
+  return [saving, save];
+}
+
+function getDefaultTitle(text) {
+  return capitalize(words(text).join(' ')); // humanize
+}
+
+function useNewName(existingParams) {
+  const [name, setName] = useState('');
+  const [feedback, setFeedback] = useState({ isError: false, text: null });
+
+  useEffect(() => {
+    const alreadyExists = includes(existingParams, name);
+    if (alreadyExists) {
+      setFeedback({ isError: true, helpText: 'Parameter with this name already exists' });
+    } else if (!name) {
+      setFeedback({ isError: false, helpText: 'Choose a keyword for this parameter' });
+    } else {
+      setFeedback({ isError: false, helpText: `This is what will be added to your query editor {{ ${name} }}` });
+    }
+  }, [name]);
+
+  return [feedback, name, setName];
+}
```

Comment:
While this is an interesting use of `useEffect` and a custom hook, I think a small component will result in a more readable code:

https://codesandbox.io/s/3yj38pv68q

In the example above `App` is your `EditParameterSettingsDialog`.

Reply:
🤔 doesn't Ant have a facility of their own to disable on errors?

Reply:
Yes but I don't want to use. It's code mess like you won't believe.

Reply:
I guess we can introduce some callback from the input to the parent component, although this starts to become messy too.

The code mess is what you discussed with Gabriel?

Reply:
It's beyond the scope of this PR, so here let's resort to less ideal workarounds. But for the long term we should either embrace Ant's mess, implement some abstraction on top of it and implement our own abstraction. One way or another, no point in repeating this in every form, right?

Reply:
I built this form in a way that has only one "feedback" field to avoid complexity.
I do favor your component approach, much cleaner and more readable. I'll see if I can work around it.

Reply:
👍 I updated the CodeSandbox to include a possible solution for this. It's not ideal, but...

Reply:
Ha! I'm working on the same solution.
Problem is that it works only if one component is allowed to set validity.

Reply:
Oh, right. New option, that actually cleaner but introduces mild code duplication:

```javascript
function okDisabled () {
  return isNew && (includes(existingsParams, param.name) || !param.name) || param.title === '';
}
```

We dup the existing params check, but we don't need to pass around onError.

Reply:
@arikfr You can create a function to validate name, use it in dialog and pass as prop to the name input component

Reply:
@arikfr this works for me
https://codesandbox.io/s/ry3q97l83p

Reply:
(updated PR)

Reply:
@kravets-levko that's an option as well.

@ranbena 👍

Reply:
> just implement our own abstraction

:+1: 

I'm not a very fan of Ant Form coding style either and I wouldn't say it's worth it when you have only one field with validation.